src/ConnectedMode.UnitTests/ServerSentEvents/ServerSentEventPumpTests.cs

```diff
@@ -0,0 +1,116 @@
+
+namespace SonarLint.VisualStudio.ConnectedMode.UnitTests.ServerSentEvents
+{
+    [TestClass]
+    public class ServerSentEventPumpTests
+    {
+        private Mock<IServerSentEventsSession> serverSentEventsSessionMock;
+        private Mock<IIssueChangedServerEventSourcePublisher> issueChangedPublisherMock;
+        private Mock<ITaintServerEventSourcePublisher> taintPublisherMock;
+        private ServerSentEventPump testSubject;
+
+        [TestInitialize]
+        public void SetUp()
+        {
+            serverSentEventsSessionMock = new Mock<IServerSentEventsSession>();
+            issueChangedPublisherMock = new Mock<IIssueChangedServerEventSourcePublisher>();
+            taintPublisherMock = new Mock<ITaintServerEventSourcePublisher>();
+
+            testSubject = new ServerSentEventPump(issueChangedPublisherMock.Object, taintPublisherMock.Object);
+        }
+
+        [TestMethod]
+        public void Dispose_DisposesPublishers()
+        {
+            testSubject.Dispose();
+
+            issueChangedPublisherMock.Verify(ch => ch.Dispose(), Times.Once);
+            taintPublisherMock.Verify(ch => ch.Dispose(), Times.Once);
+        }
+
+        [TestMethod]
+        public void MefCtor_CheckIsExported()
+        {
+            MefTestHelpers.CheckTypeCanBeImported<ServerSentEventPump, IServerSentEventPump>(
+                MefTestHelpers.CreateExport<IIssueChangedServerEventSourcePublisher>(),
+                MefTestHelpers.CreateExport<ITaintServerEventSourcePublisher>());
+        }
+
+        [TestMethod]
+        public async Task PumpAllAsync_SelectsPublisherCorrectlyAndPreservesOrderWithinType()
+        {
+            var inputSequence = new IServerEvent[]
+            {
+                Mock.Of<IIssueChangedServerEvent>(),
+                Mock.Of<ITaintVulnerabilityRaisedServerEvent>(),
+                Mock.Of<IIssueChangedServerEvent>(),
+                Mock.Of<ITaintVulnerabilityClosedServerEvent>(),
+                Mock.Of<ITaintVulnerabilityClosedServerEvent>(),
+                Mock.Of<IIssueChangedServerEvent>(),
+                Mock.Of<ITaintVulnerabilityRaisedServerEvent>(),
+            };
+            SetUpEventsSequenceThatFinishes(inputSequence);
+
+            await testSubject.PumpAllAsync(serverSentEventsSessionMock.Object);
+            
+            issueChangedPublisherMock.Invocations
+                .Select(call => call.Arguments.First() as IIssueChangedServerEvent)
+                .Should()
+                .BeEquivalentTo(inputSequence.Where(issuesEvent => issuesEvent is IIssueChangedServerEvent));
+            taintPublisherMock.Invocations
+                .Select(call => call.Arguments.First() as ITaintServerEvent)
+                .Should()
+                .BeEquivalentTo(inputSequence.Where(taintEvent => taintEvent is ITaintServerEvent));
+        }
+
+        [TestMethod]
+        public async Task PumpAllAsync_WhenPublisherDisposed_Finishes()
```

Comment:
This test doesn't logically make sense now that the pump is the one disposing.
Instead, there should be a test for what happens if you call PumpAll after disposing the pump.